Oceanwater 949 - ArmFindSurface

AstroStucky opened the pull request:
## Linked Issues:
| EPIC ⚡ | [OCEANWATER-933](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-933) |
| :----------- | :----------- |
| Jira Ticket 🎟️ | [OCEANWATER-949](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-949) |

## Summary of Changes
* **ArmFindSurface** action added
* Fixed some potential bugs in mixins.py
* tf2 for Python appears to be missing 3D math operations for geometry_msgs types that are otherwise available in C++. A math3d module has been created to handle these kind of operations.
* Deprecated GuardedMove. 
  * _Should it also print a deprecated warning when used?_

## Related Branches
Your **ow_autonomy** repo should be on the epic branch, `OCEANWATER-933-command-unification-epic`, when testing.

## Tests
### Help Message
Try it out and make sure the language makes sense.
```
rosrun ow_lander arm_find_surface.py --help
```

### Poke Terrain with Absolute Commands
Launch atacama_y1a and perform the following 
#### 1. Unstow the arm. This is good to do before any operation that poses the arm anywhere near the terrian.
`rosrun ow_lander arm_unstow.py`
Testing will focus on this reddish mound near the lander.
![default_gzclient_camera(1)-2023-01-18T12_32_15 610637](https://user-images.githubusercontent.com/17039973/213276982-92a92e0f-d7e7-41b0-8fdf-28abd306234d.jpg)

#### 2. Fail to find the mound from above
`rosrun ow_lander arm_find_surface.py -x 1.45 -y -0.4 -z 0.15`
Scoop will stop just above it and you will see 
```
ArmFindSurface action started
ArmFindSurface: Succeded - No surface was found
ArmFindSurface action complete
```
![default_gzclient_camera(1)-2023-01-18T13_14_57 884971](https://user-images.githubusercontent.com/17039973/213285203-326e05d5-5a06-4f18-b675-e5d66924b1d8.jpg)

#### 3. Find the mound from above
`rosrun ow_lander arm_find_surface.py -x 1.45 -y -0.4`
You will see the following 
```
ArmFindSurface action started
ArmFindSurface: Succeded - ArmFindSurface trajectory stopped by a force of 379.18 N. Surface found at (1.460, -0.402, -0.051)
ArmFindSurface action complete
```
![default_gzclient_camera(1)-2023-01-18T12_34_22 730745](https://user-images.githubusercontent.com/17039973/213277319-511a490a-d5f6-48b2-8b23-dd257776da41.jpg)

#### 4. Find a side slope of the mound
Retract off the surface by 8 cm to reset for another poke
`rosrun ow_lander arm_move_cartesian.py -r -z -0.2`

Attempt to find the side slope of the mound
`rosrun ow_lander arm_find_surface.py -x 1.45 -y -.36 -z -.05 -n 0 -.5 -.5`
You should see
```
ArmFindSurface action started
ArmFindSurface: Succeded - ArmFindSurface trajectory stopped by a force of 229.76 N. Surface found at (1.452, -0.377, -0.064)
ArmFindSurface action complete
```
![default_gzclient_camera(1)-2023-01-18T12_46_25 241832](https://user-images.githubusercontent.com/17039973/213279916-8a3be81c-f2d3-451c-a1d0-0d14b2d28fe0.jpg)

### Poke Terrain with Relative Commands
You can do the following in the same simulation instance
#### 1. Interrogate the flat spot in front of the lander
`rosrun ow_lander arm_find_surface.py -x 1.6 -y 0.1 -z -0.1`
```
[INFO:/lander_action_servers] ArmFindSurface action started
[INFO:/lander_action_servers] ArmFindSurface: Succeded - ArmFindSurface trajectory stopped by a force of 450.79 N. Surface found at (1.604, 0.098, -0.178)
[INFO:/lander_action_servers] ArmFindSurface action complete
```
![default_gzclient_camera(1)-2023-01-18T12_52_03 570250](https://user-images.githubusercontent.com/17039973/213280721-7d8bcb38-dc30-4266-969f-cff339618797.jpg)

#### 2. Interrogate multiple points extending out from the side of the scoop by repeating the the following 2 commands
1. Reset for next poke `rosrun ow_lander arm_move_cartesian.py -r -z -0.06`
2. Poke 10 cm to the scoop's left `rosrun ow_lander arm_find_surface.py -r -y -0.1`
3. Repeat the previous 2 steps as many times as you'd like.

The following output shows my results after performing these steps a total of 4 times
```
ArmMoveCartesian action complete
ArmFindSurface action started
ArmFindSurface: Succeded - ArmFindSurface trajectory stopped by a force of 245.79 N. Surface found at (1.613, 0.199, -0.177)
ArmFindSurface action complete
ArmMoveCartesian action started
ArmMoveCartesian: Succeded - ArmMoveCartesian trajectory succeeded
ArmMoveCartesian action complete
ArmFindSurface action started
ArmFindSurface: Succeded - ArmFindSurface trajectory stopped by a force of 283.55 N. Surface found at (1.626, 0.293, -0.200)
ArmFindSurface action complete
ArmMoveCartesian action started
ArmMoveCartesian: Succeded - ArmMoveCartesian trajectory succeeded
ArmMoveCartesian action complete
ArmFindSurface action started
ArmFindSurface: Succeded - ArmFindSurface trajectory stopped by a force of 292.45 N. Surface found at (1.639, 0.401, -0.204)
ArmFindSurface action complete
ArmMoveCartesian action started
ArmMoveCartesian: Succeded - ArmMoveCartesian trajectory succeeded
ArmMoveCartesian action complete
ArmFindSurface action started
ArmFindSurface: Succeded - ArmFindSurface trajectory stopped by a force of 270.71 N. Surface found at (1.640, 0.505, -0.175)
ArmFindSurface action complete
```
![default_gzclient_camera(1)-2023-01-18T12_57_45 015485](https://user-images.githubusercontent.com/17039973/213282086-c9759c7a-5556-4214-85e6-04c7b03605f5.jpg)

kmdalal commented:
This branch appears to need the epic branch merged into it...

AstroStucky commented:
> This branch appears to need the epic branch merged into it...

Everything shows up-to-date for me. Is it possible you have unpushed changes on your version of the epic branch?

kmdalal commented:
> > This branch appears to need the epic branch merged into it...
> 
> Everything shows up-to-date for me. Is it possible you have unpushed changes on your version of the epic branch?

Upon checking out your branch and doing a clean build of the workspace, I had a compile error in OwInterface.h.  I did a fast-forward merge from the epic branch OW-933 (which should have just said "Everything is up to date" if it wasn't already merged).   But now everything is fine.

kmdalal commented:
I deleted my last comment, which was premature.  This branch is still not building for me; I'm using noetic-devel of ow_autonomy.
```
/usr/local/home/kdalal/ow/src/ow_autonomy/ow_plexil/src/plexil-adapter/OwInterface.h:27:10: fatal error: ow_lander/UnstowAction.h: No such file or directory
   27 | #include <ow_lander/UnstowAction.h>
```

AstroStucky commented:
> I deleted my last comment, which was premature. This branch is still not building for me; I'm using noetic-devel of ow_autonomy.
> 
> ```
> /usr/local/home/kdalal/ow/src/ow_autonomy/ow_plexil/src/plexil-adapter/OwInterface.h:27:10: fatal error: ow_lander/UnstowAction.h: No such file or directory
>    27 | #include <ow_lander/UnstowAction.h>
> ```

You do need to test this with **ow_autonomy** on `OCEANWATER-933-command-unification-epic`. Otherwise the rename of stow and unstow will break your build. 

I just updated the original post to mention this.

kmdalal commented:
I get compile errors in ow_plexil whether using the OCEANWATER-933-command-unification-epic branch or the OCEANWATER-934_telemetry_unification_epic branches.  Different errors.  This seems to indicate that elements of both epic branches are needed.  I'll experiment.

BTW, I just updated both epic branches on both ow_simulator and ow_autonomy from noetic-devel and pushed these.

AstroStucky commented:
> Different errors. This seems to indicate that elements of both epic branches are needed.

I have to disagree with this assessment. `OCEANWATER-934_telemetry_unification_epic` should have nothing to do with this PR.

I have cleaned and built several times in a fresh terminal with all repo branches recently updated from Babelfish, and have not been able to replicate the **ow_plexil** compile error you shared earlier.

kmdalal commented on ow_lander/scripts/arm_find_surface.py:
```diff
@@ -0,0 +1,58 @@
+
+parser = argparse.ArgumentParser(
+  formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+  description="Determine the position of a surface by moving the arm's " \
+              "end-effector towards surface until sufficient resistance is " \
+              "encountered.")
```
Minor: I'm not sure the backslashes are needed.  I've omitted them in another development with no problems.

AstroStucky replied:
Looks like I have definitely been using these line splits where they are not needed. They have been removed from this file, but I will save a full purge of the unnecessary ones for later so not to clutter this branch with minor changes. It will probably be done as a commit on noetic-devel.

kmdalal commented:
> I have cleaned and built several times in a fresh terminal with all repo branches recently updated from Babelfish, and have not been able to replicate the **ow_plexil** compile error you shared earlier.

Thanks for the effort.  I deleted this branch and the command epic branch in my ow_autonomy clone, refetched them, and everything built.  Not sure what was the problem in my previous environment.

AstroStucky commented:
I'm glad it's working now, Mike. It is unfortunate the resolution does not shed much light on what was going wrong.

kmdalal commented:
My first attempt at test step 2 failed during planning:
```
[INFO]: ArmFindSurface action started
[ERROR]: FrameTransfomer.transform failure: "world" passed to lookupTransform argument target_frame does not exist. 
[ERROR]: ArmFindSurface: Aborted - Trajectory planning failed - Setup trajectory failed
[INFO]: ArmFindSurface action complete
```

AstroStucky commented:
Hmmm. 🤔 This is the first I have seen this error. Strange too, because Gazebo most certainly defines a "world" frame. I will add a note comment the relevant FrameTransformer method and try and keep an eye on this.

kmdalal commented:
Step 3 succeeded, though to my surprise it deformed the terrain a bit.  The "dent" made by the scoop is visible in your snapshot.   Either the material is softer or the arm more forceful than I'd have guessed.

AstroStucky commented:
> Step 3 succeeded, though to my surprise it deformed the terrain a bit. The "dent" made by the scoop is visible in your snapshot. Either the material is softer or the arm more forceful than I'd have guessed.

Deforming does happen. Whenever the scoop points down, it is capable of deforming the terrain. It does not matter the "softness" because that concept does not really exist in our current terrain deformation model. This was actually already happening with GuardedMove for a long time now. I think the deformations appear to be more extreme when an inclined poke is used like in **Step 4**.

kmdalal reviewed (approved):
Tests pass, with one anomaly noted and acted on.   Much detail and clarity in the client script's help message.

Code looks clean, albeit my fairly light review.  I like the new math library.

Great work!  This is a very useful addition.  When the branch is merged, I (or you) can replace the GuardedMoves in PLEXIL.  Some interesting terramechanical inspection plans are now possible.

AstroStucky commented on ow_lander/scripts/arm_find_surface.py:
```diff
@@ -0,0 +1,58 @@
+       "be interpreted in the base_link frame.")
+parser.add_argument('--distance', '-d', type=float, default=0.1,
+  help="Distance away from the estimated surface position that the " \
+       "end-effector start its trajectory.")
```
fixed

AstroStucky commented:
Actually scraping of the surface during the pre-poke movement for **Step 4** happens pretty often when I try it, so my previous comment may have been mistaken. Increasing the default **distance** to 0.3 did not really solve it, so I think it's really more a test procedure issue than a parameter one, and this might be important to keep in mind when we write autonomy plans that interrogate the surface like this. I have changed the test procedure so that the back-off before the arm find surface in **Step 4** is 20cm instead of 8cm.

AstroStucky commented:
I'll also increase the default **distance** to 20 cm.
